src/DotNetWorker/Hosting/FunctionsApplicationBuilder.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Azure.Functions.Worker.Hosting;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.Metrics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Extensions.Hosting;
+
+/// <summary>
+/// A builder for Azure Functions worker applications.
+/// </summary>
+public class FunctionsApplicationBuilder : IHostApplicationBuilder, IFunctionsWorkerApplicationBuilder
+{
+    private readonly HostApplicationBuilder _hostApplicationBuilder;
+    private readonly BootstrapHostBuilder _bootstrapHostBuilder;
+    private readonly IFunctionsWorkerApplicationBuilder _functionsWorkerApplicationBuilder;
+
+    private bool _hasBuilt = false;
+
+    internal const string SkipDefaultWorkerMiddlewareKey = "__SkipDefaultWorkerMiddleware";
+
+    internal FunctionsApplicationBuilder(string[]? args)
+    {
+        var configuration = new ConfigurationManager();
+
+        _hostApplicationBuilder = new HostApplicationBuilder(new HostApplicationBuilderSettings
+        {
+            Configuration = configuration
+        });
+
+        _bootstrapHostBuilder = new BootstrapHostBuilder(_hostApplicationBuilder);
+
+        // We'll set the default middleware up just before we build.
+        _bootstrapHostBuilder.Context.Properties[SkipDefaultWorkerMiddlewareKey] = true;
+
+        _bootstrapHostBuilder
+                   .ConfigureDefaults(args)
+                   .ConfigureFunctionsWorkerDefaults();
+
+        _functionsWorkerApplicationBuilder = InitializeHosting(_bootstrapHostBuilder);
+    }
+
+    /// <inheritdoc/>
+    public IDictionary<object, object> Properties => ((IHostApplicationBuilder)_hostApplicationBuilder).Properties;
+
+    IConfigurationManager IHostApplicationBuilder.Configuration => Configuration;
+
+    /// <inheritdoc/>
+    public ConfigurationManager Configuration => _hostApplicationBuilder.Configuration;
+
+    /// <inheritdoc/>
+    public IHostEnvironment Environment { get; private set; }
+
+    /// <inheritdoc/>
+    public ILoggingBuilder Logging => _hostApplicationBuilder.Logging;
+
+    /// <inheritdoc/>
+    public IMetricsBuilder Metrics => _hostApplicationBuilder.Metrics;
+
+    /// <inheritdoc />
+    public IServiceCollection Services => _hostApplicationBuilder.Services;
```

Comment:
You may also have an issue with `/// <inheritdoc />` here as both interfaces expose this property.

Reply:
that one seems to work...